gobblin-core/src/main/java/gobblin/publisher/TimePartitionedDataWithRecordCountsPublisher.java

```diff
@@ -0,0 +1,150 @@
+ * directories at a time. This is done so that each task/branch can move the file it created,
+ * allowing it to access the row count information and place it into the file name.
+ *
+ * Publisher output directory is $GOBBLIN_WORK_DIR/job-output/{tableName}/{partitionPath}
+ *
+ * @author ekrogen
+ */
+public class TimePartitionedDataWithRecordCountsPublisher extends TimePartitionedDataPublisher {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TimePartitionedDataWithRecordCountsPublisher.class);
+
+  public TimePartitionedDataWithRecordCountsPublisher(State state) {
+    super(state);
+  }
+
+  @Override
+  public void publishData(Collection<? extends WorkUnitState> states) throws IOException {
+
+    Set<Path> publisherDirsProcessed = Sets.newHashSet();
+
+    for (WorkUnitState workUnitState : states) {
+      for (int branchId = 0; branchId < this.numBranches; branchId++) {
+
+        Path writerOutputDir = WriterUtils.getWriterOutputDir(workUnitState, this.numBranches, branchId);
+
+        FileSystem fileSystem = this.fss.get(branchId);
+        if (!fileSystem.exists(writerOutputDir)) {
+          LOG.warn("WorkUnit " + workUnitState.getId() + " produced no data");
+          workUnitState.setWorkingState(WorkUnitState.WorkingState.COMMITTED);
+          break;
+        }
+
+        boolean replaceFinalOutputDir = this.getState().getPropAsBoolean(ForkOperatorUtils
+            .getPropertyNameForBranch(ConfigurationKeys.DATA_PUBLISHER_REPLACE_FINAL_DIR, this.numBranches, branchId));
+
+        // The directory where the final output directory for this job will be placed. It is a combination of
+        // DATA_PUBLISHER_FINAL_DIR and WRITER_FILE_PATH
+        Path publisherOutputDir = WriterUtils.getDataPublisherFinalDir(workUnitState, this.numBranches, branchId);
+
+        if (!publisherDirsProcessed.contains(publisherOutputDir) && replaceFinalOutputDir
+            && fileSystem.exists(publisherOutputDir)) {
+          // If the final output directory is configured to be replaced, and we haven't yet
+          // processed (deleted) it, and it exists, then delete it to begin replacing it
+          fileSystem.delete(publisherOutputDir, true);
+        }
+        publisherDirsProcessed.add(publisherOutputDir);
+
+        List<String> filePathToRecordCountMappings = workUnitState.getPropAsList(ForkOperatorUtils
+            .getPropertyNameForBranch("paths.to.record.counts.mapping", this.numBranches, branchId), "");
+
+        if (filePathToRecordCountMappings.isEmpty()) {
+          LOG.warn("No record count mappings found for WorkUnit " + workUnitState.getId() + " branch " + branchId);
+        }
+        for (String filePathToRecordCountMap : filePathToRecordCountMappings) {
+          String filePath;
+          int recordsWritten;
+          try {
+            filePath = filePathToRecordCountMap.split(":")[0];
+            recordsWritten = Integer.parseInt(filePathToRecordCountMap.split(":")[1]);
```

Comment:
Let me know if you have any suggestions on a better way to achieve this type of mapping - thought about JSON serializing but that seemed like overkill